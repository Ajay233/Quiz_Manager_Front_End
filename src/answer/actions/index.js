import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
