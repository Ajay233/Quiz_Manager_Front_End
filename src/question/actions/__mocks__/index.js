export const getQuestions = jest.fn()

export const setCurrentQuestion = jest.fn()

export const deleteQuestion = jest.fn()

export const addQuestion = jest.fn()

export const updateQuestion = jest.fn()

export const clearQuestions = jest.fn()
