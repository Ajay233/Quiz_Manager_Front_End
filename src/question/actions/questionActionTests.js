import configureMockStore from 'reduc-mock-store'
import thunk from 'redux-thunk'
import {
  getQuestions,
  setCurrentQuestion,
  deleteQuestion,
  addQuestion,
  updateQuestion,
  clearQuestions
} from './index'

jest.mock("../../utils/session")
jest.mock("../../axiosRequests/axiosUtil")

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe("getQuestions", () => {
  it("should return an action that will add a list of questions the questions store", () => {
    const store = mockStore({})

    const expectedAction = {
      type: "SET_QUESTIONS",
      payload: [{ id: 1, questionNumber: 1 }, { id: 2, questionNumber: 2 }, { id: 3, questionNumber: 3 }]
    }

    return store.dispatch(getQuestions("question/findByQuizId", "data", "jwt")).then(() => {
      expect(store.getActions()[0]).toEqual(expectedAction)
    })
  })

  it("should call sessionExpired if an error occurrs and the status id 403", () => {
    const store = mockStore({})

    return store.dispatch(getQuestions("sessionExpired", "data", "Jwt")).then(() => {
      expect(sessionExpired).toHaveBeenCalledTimes(1);
    })
  })
})

describe("deleteQuestion", () => {
  it("should return an action that will delete a question from the questions store", () => {
    const expectedAction = {
      type: "DELETE_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(deleteQuestion(question)).toEqual(expectedAction)
  })
})

describe("addQuestion", () => {
  it("should return an action that will add a question to the questions store", () => {
    const expectedAction = {
      type: "ADD_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(addQuestion(question)).toEqual(expectedAction)
  })
})

describe("updateQuestion", () => {
  it("should return an action that will update a question in the questions store", () => {
    const expectedAction = {
      ype: "UPDATE_QUESTION",
      payload: { id: 1 }
    }

    expect(updateQuestion(question)).toEqual(expectedAction)
  })
})

describe("clearQuestions", () => {
  it("should return an action that will clear the questions store", () => {
    const expectedAction = {
      type: "CLEAR_QUESTIONS",
      payload: []
    }

    expect(clearQuestions()).toEqual(expectedAction)
  })
})

describe("setCurrentQuestion", () => {
  it("should return an action that will add a question the question store", () => {
    const expectedAction = {
      type: "SET_CURRENT_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(setCurrentQuestion(question)).toEqual(expectedAction)
  })
})
