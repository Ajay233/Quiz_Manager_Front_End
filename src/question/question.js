import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

class Question extends React.Component {

  renderQuestion = () => {
    const { question_number, description } = this.props.question
    return(
      <div>
        <div className="questionNumber">{question_number}</div>
        <div className="questionDescription">{description}</div>
        <div className="questionOptions">{this.renderOptions()}</div>
      </div>
    )
  }

  renderOptions = () => {
    const { permission } = this.props.userData;
    return(
      <div className="options">
        { permission === "USER" ? <Link to="#" className="start" onClick={this.handleStart}><i className="far fa-play-circle blue"></i> Start</Link> : null }
        { permission === "READ-ONLY" ? <Link to="#" className="view" onClick={this.handleView}><i className="far fa-eye blue"></i> View</Link> : null }
        { permission === "ADMIN" ? <Link to="#" className="edit" onClick={this.handleEdit}><i className="fas fa-edit blue"></i> Edit</Link> : null }
        { permission === "ADMIN" ? <Link to="#" className="deleteOption" onClick={this.handleDelete}><i className="fas fa-trash-alt red"></i> Delete</Link> : null }
      </div>
    );
  }

  render(){
    return(
      <div>
        {this.renderQuestion()}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    userData: state.userData
  }
}

export default connect(mapStateToProps)(Question)
