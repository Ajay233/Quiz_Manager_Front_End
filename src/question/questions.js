import React from 'react';
import Question from './question'

const Questions = (props) => {

  const renderQuestions = () => {
    const { questions, userData, getAnswers, setCurrentQuestion, setNotification, deleteQuestion, showModal } = props
    const listOfQuestions = questions.map(question => {
      return( <Question
                key={questions.indexOf(question)}
                question={question}
                userData={userData}
                getAnswers={getAnswers}
                setCurrentQuestion={setCurrentQuestion}
                setNotification={setNotification}
                deleteQuestion={deleteQuestion}
                showModal={showModal}
              />
      );
    });
    return listOfQuestions;
  }

  return(
    <div>
      {renderQuestions()}
    </div>
  );
  
}

export default Questions
