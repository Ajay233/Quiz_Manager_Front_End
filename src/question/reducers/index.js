import { questions } from '../factory/questionsFactory'

export const setQuestionsReducer = (state=questions, action) => {
  switch (action.type) {
    case "SET_QUESTIONS": return action.payload;
    case "DELETE_QUESTION": return state.filter(question => question!== action.payload)
    default: return state;
  }
}


export const setCurrentQuestionReducer = (state={}, action) => {
  switch (action.type) {
    case "SET_CURRENT_QUESTION": return action.payload
    default: return state;
  }
}
