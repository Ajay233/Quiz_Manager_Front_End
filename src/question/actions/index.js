import { reset } from 'redux-form'
import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { getAnswers } from '../../answer/actions'
import { sessionExpired } from '../../utils/session'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { setLoaderState } from '../../components/actions'
import history from '../../history'

export const addQuestion = (body, jwt) => {
  return (dispatch) => {
    dispatch(setLoaderState(true, "Saving...", "saving"))
    return post("question/create", body, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch({
        type: "ADD_QUESTION",
        payload: response.data
      })
      history.push("/editQuiz");
      dispatch(setNotification("Question created", "success", true));
    }).catch((error) => {
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.status === 500 && error.response.data.message.includes("Maximum upload size exceeded")) {
        dispatch(setNotification("Error - Image exceeds the maximum size of 1040KB", "error", true));
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create question with the details provided", "error", true))
      }
    });
  }
}

export const getQuestions = (endpoint, param, jwt, startQuiz=false) => {
  return (dispatch) => {
    return getUsingParams(endpoint, param, jwt).then((response) => {
      if(response.data !== "NO QUESTIONS"){
        dispatch({
          type: "SET_QUESTIONS",
          payload: response.data
        })
      } else {
        dispatch({
          type: "SET_QUESTIONS",
          payload: []
        })
      }
      if(startQuiz){
        const body = { questionId: response.data[0].id }
        dispatch(getAnswers(body, jwt))
        dispatch(setCurrentQuestion(response.data[0]));
      }
      console.log("success")
    }).catch((error) => {
      console.log(error);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification(error.response.data, "error", true))
      }
    });
  }
}

export const updateQuestion = (body, jwt) => {
  return (dispatch) => {
    dispatch(setLoaderState(true, "Saving...", "saving"))
    return put("question/update", body, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch(setCurrentQuestion(response.data));
      dispatch({
        type: "UPDATE_QUESTION",
        payload: response.data
      })
      dispatch(setNotification("Question updated", "success", true))
      dispatch(reset('updateQuestionForm'))
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.status === 500 && error.response.data.message.includes("Maximum upload size exceeded")) {
        dispatch(setNotification("Error - Image exceeds the maximum size of 1040KB", "error", true));
      } else {
        dispatch(setNotification("Error - unable to update question", "error", true))
      }
    });
  }
}

export const setCurrentQuestion = (question) => {
  return {
    type: "SET_CURRENT_QUESTION",
    payload: question
  }
}

export const deleteQuestion = (config, jwt) => {
  return (dispatch) => {
    dispatch(hideModal())
    dispatch(setLoaderState(true, "Deleting..."))
    return del("question/delete", config, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch({
        type: "DELETE_QUESTION",
        payload: config.data[0]
      })
      history.push("/editQuiz")
      dispatch(setNotification("Question deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - Unable to delete this question", "error", true))
      }
    });
  }
}

export const deleteImage = (config, jwt) => {
  return (dispatch) => {
    dispatch(hideModal())
    dispatch(setLoaderState(true, "Deleting..."))
    return del("question/deleteImage", config, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch(setCurrentQuestion(response.data))
      dispatch({
        type: "UPDATE_QUESTION",
        payload: response.data
      })
      dispatch(setNotification("Image has been deleted", "success", true))
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - Unable to delete this image", "error", true))
      }
    })
  }
}

export const clearQuestions = () => {
  return {
    type: "CLEAR_QUESTIONS",
    payload: []
  }
}
