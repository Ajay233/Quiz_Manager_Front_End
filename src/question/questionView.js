import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Answers } from '../answer/answers'
import UpdateQuestionForm from './forms/updateQuestion'
import Notification from '../notifications/notifications'
import Modal from '../modal/modal'
import FilePlaceholder from '../components/filePlaceholder'
import history from '../history'

import { setNotification } from '../notifications/actions'
import { hideModal, showModal, showModal2, showModal3 } from '../modal/actions'
import { setCurrentAnswer, deleteAnswer } from '../answer/actions'
import { deleteQuestion, deleteImage } from '../question/actions'

import '../stylesheets/question.css'
import '../stylesheets/answer.css'

export class QuestionView extends React.Component {

  componentDidMount(){
    document.documentElement.scrollTop = 0;
  }

  renderModal = () => {
    const { currentAnswer, hideModal, currentQuestion } = this.props
    const { showModal, showModal2, showModal3 } = this.props.modalState
    if(showModal === true){
      return(
        <Modal
          type={"delete"}
          show={showModal}
          title={"Delete Answer"}
          message={`You are about to delete answer ${currentAnswer.answerIndex}.  Do you want to continue?`}
          onDelete={this.handleDeleteAnswer}
          onCancel={hideModal}
        />
      );
    } else if(showModal2 === true){
      return(
        <Modal
          type={"delete"}
          show={showModal2}
          title={"Delete Question"}
          message={`You are about to delete question ${currentQuestion.questionNumber}, this will also delete any associated answers for this question.  Do you want to continue?`}
          onDelete={this.handleDeleteQuestion}
          onCancel={hideModal}
        />
      );
    } else if(showModal3 === true){
      return(
        <Modal
          show={showModal3}
          title={"Delete Image"}
          message={`You are about to delete the image for this question.  Do you want to continue?`}
          onDelete={this.handleDeleteImg}
          onCancel={hideModal}
        />
      );
    } else {
      return null
    }
  }

  renderAnswers = () => {
    const { answers, userData, setCurrentAnswer, setNotification, deleteAnswer, showModal } = this.props;
    return answers.length === 0 ? null :
    <Answers
      answers={answers}
      userData={userData}
      setCurrentAnswer={setCurrentAnswer}
      setNotification={setNotification}
      deleteAnswer={deleteAnswer}
      showModal={showModal}
    />
  }

  renderDetails = () => {
    const { currentQuestion, quiz } = this.props
    return(
      <React.Fragment>
        <div className="questionViewTitle">{`Question ${currentQuestion.questionNumber} from the ${quiz.name} quiz`}</div>
        <div className={"questionDetailsContainerLarge"}>
          <div className="detailsHeading">Question description:</div>
          <div className="detailsContent">{ currentQuestion.description }</div>
        </div>
      </React.Fragment>
    );
  }

  renderImg = () => {
    const { showModal3, userData } = this.props
    return(
      <div className="questionImageArea">
        {this.renderQuestionImage()}
        <div className="link deleteImg" onClick={() => showModal3()}>
          {userData.permission === "ADMIN" || userData.permission === "SUPER-USER" ? <span><i className="fas fa-trash-alt red"></i> Delete Image</span> : null}
        </div>
      </div>
    );
  }

  renderFormOrDetails = () => {
    const { currentQuestion, userData } = this.props
    if(history.location.pathname === "/editQuestion"){
      return(
        <div className={currentQuestion.imgUrl !== null ? 'questionFormArea' : ''}>
          <div className={currentQuestion.imgUrl !== null ? 'questionForm' : ''}>
            <UpdateQuestionForm triggerModal={this.triggerModal}/>
          </div>
          {currentQuestion.imgUrl !== null ? this.renderImg() : null}
        </div>
      );
    } else {
      return(
        <div className={'questionFormArea'}>
          <div className={'questionForm'}>
            {this.renderDetails()}
          </div>
          {currentQuestion.imgUrl !== null ? this.renderImg() : null}
          {this.renderPlaceholder(currentQuestion.imgUrl, userData.permission)}
        </div>
      );
    }
  }

  renderPlaceholder = (imgUrl, permission) => {
    if(imgUrl === null && permission !== "ADMIN" && permission !== "SUPER-USER"){
      return(
        <div className="questionImageArea">
        <FilePlaceholder type={'question'} permission={permission}/>
        </div>
      );
    } else {
      return null
    }
  }

  renderQuestionImage = () => {
    const { currentQuestion } = this.props
    return currentQuestion.imgUrl || currentQuestion.imgUrl !== null ? <div className="imgView"><img src={currentQuestion.imgUrl} alt="" className="preview"/></div> : null
  }

  handleDeleteAnswer = () => {
    const { currentAnswer, userData, deleteAnswer } = this.props;
    const config = {
      data: [currentAnswer]
    }
    deleteAnswer(config, userData.jwt)
  }

  handleDeleteQuestion = () => {
    const { currentQuestion, userData, deleteQuestion } = this.props;
    const config = {
      data: [currentQuestion]
    }
    deleteQuestion(config, userData.jwt)
  }

  handleDeleteImg = () => {
    const { currentQuestion, userData, deleteImage } = this.props;
    const config = {
      params: {
        questionId: currentQuestion.id,
        url: currentQuestion.imgUrl
      }
    }
    deleteImage(config, userData.jwt)
  }

  clearNotification = () => {
    const { setNotification } = this.props
    setNotification()
  }

  renderAddButton = () => {
    const { permission } = this.props.userData
    return permission === "ADMIN" || permission === "SUPER-USER" ? <Link to="/newAnswer" className="addButton linkStandard" onClick={this.clearNotification}><i className="fas fa-plus-circle green"></i> Add an answer</Link> : null
  }

  triggerModal = (event) => {
    event.preventDefault()
    this.props.showModal2()
  }

  renderBackButton = () => {
    const { permission } = this.props.userData
    if(permission === "READ-ONLY"){
      return <Link to="/viewQuiz" className="link back" onClick={this.clearNotification}><i className="fas fa-chevron-circle-left blue"></i> Back</Link>
    } else {
      return <Link to="/editQuiz" className="link back" onClick={this.clearNotification}><i className="fas fa-chevron-circle-left blue"></i> Back</Link>
    }
  }

  render(){
    return(
      <div id="questionView" className="componentContainer">
        {this.renderModal()}
        {this.renderBackButton()}
        <Notification />
        {this.renderFormOrDetails()}
        <div className="headerContainer">
          <div id="answersTitle">Answers</div>
          {this.renderAddButton()}
        </div>
        <div id="answerHeadings" className="answerHeadings">
          <div id="numberHeader">Answer</div>
          <div id="descriptionHeader">Description</div>
          <div id="correctAnswerHeading">Correct answer?</div>
          <div id="optionsHeader"></div>
        </div>
        <div id="answers">
          {this.renderAnswers()}
        </div>
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    quiz: state.quiz,
    currentQuestion: state.currentQuestion,
    answers: state.answers,
    currentAnswer: state.currentAnswer,
    modalState: state.modalState
  }
}

export default connect(mapStateToProps,
  {
    setNotification,
    hideModal,
    deleteAnswer,
    setCurrentAnswer,
    deleteQuestion,
    deleteImage,
    showModal,
    showModal2,
    showModal3
  })(QuestionView)
