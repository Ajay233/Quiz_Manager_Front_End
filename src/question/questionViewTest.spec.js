import React from 'react'
import Modal from '../modal/modal'
import ReactDOM from 'react-dom'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import { QuestionView } from './questionView'
import { mapStateToProps } from './questionView'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { render, fireEvent, cleanup } from '@testing-library/react'
import history from '../history'

import { setNotification } from '../notifications/actions'
import { hideModal, showModal, showModal2 } from '../modal/actions'
import { setCurrentAnswer, deleteAnswer } from '../answer/actions'
import { deleteQuestion } from '../question/actions'
import { sessionExpired } from '../utils/session'

jest.mock('../notifications/actions')
jest.mock('../modal/actions')
jest.mock('../answer/actions')
jest.mock('../question/actions')
jest.mock('../utils/session')
jest.mock('../history')

const mockStore = configureStore({})

describe("mapStateToProps", () => {
  it("should map the state to props", () => {
    const appState = {
      userData: { id: 1, jwt: "jwt"},
      quiz: { id: 1, name: "test"},
      currentQuestion: { id: 1, questionNumber: 1},
      answers: [{id: 1, answerIndex: 1}, {id: 2, answerIndex: 2}],
      currentAnswer: {id: 1, answerIndex: 1},
      modalState: { showModal: false }
    }

    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("QuestionView", () => {

  let store;
  let modalState;
  let userData;
  let notificationData;
  let quiz;
  let currentQuestion;
  let answers;
  let currentAnswer;
  let component;
  let globals;

  beforeEach(() => {
    modalState = {
      showModal: false,
      showModal2: false,
      showModal3: false
    }

    userData = {
      forename: "testForename",
      surname: "testSurname",
      email: "testEmail",
      permission: "ADMIN"
    }

    notificationData = {
      message: "",
      type: "",
      show: false,
      timed: true
    }

    quiz = {
      id: 1,
      name: "Quiz 1",
      description: "Test quiz",
    }
    currentQuestion = {
      id: 1,
      questionNumber: 1,
      description: "Test question",
    }
    answers = [
      {
        id: 1,
        answerIndex: 1,
        description: "Test answer",
        correctAnswer: false
      },
      {
        id: 2,
        answerIndex: 2,
        description: "Test answer2",
        correctAnswer: true
      }
    ]

    currentAnswer = {
      id: 1,
      answerIndex: 1,
      description: "Test answer",
      correctAnswer: false
    }

    globals = { loaderState: { show: false, message: "", label: "" } }

    store = mockStore({
      userData: {
        forename: "testForename",
        surname: "testSurname",
        email: "testEmail",
        permission: "ADMIN"
      },
      notificationData:{
        message: "",
        type: "",
        show: false,
        timed: true
      }
    })

    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    })

  })

  it("should render the QuestionView component, question details and answers", () => {
    const renderedComponent = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <QuestionView
            modalState={modalState}
            userData={userData}
            notificationData={notificationData}
            currentAnswer={currentAnswer}
            currentQuestion={currentQuestion}
            quiz={quiz}
            answers={answers}
            showModal={showModal}
            showModal2={showModal2}
            hideModal={hideModal}
            setNotification={setNotification}
            deleteQuestion={deleteQuestion}
            globals={globals}
          />
        </Router>
      </Provider>
    );

    expect(renderedComponent).toMatchSnapshot()
  })

  it("should render the QuestionView component, update question form and no answers if the answers array is empty", () => {

    const answers = []

    const renderedComponent = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <QuestionView
            modalState={modalState}
            userData={userData}
            notificationData={notificationData}
            currentAnswer={currentAnswer}
            currentQuestion={currentQuestion}
            quiz={quiz}
            answers={answers}
            showModal={showModal}
            showModal2={showModal2}
            hideModal={hideModal}
            setNotification={setNotification}
            deleteQuestion={deleteQuestion}
            globals={globals}
          />
        </Router>
      </Provider>
    );

    expect(renderedComponent).toMatchSnapshot()
  })

  describe("handleDeleteQuestion", () => {

    let component

    beforeEach(() => {
      modalState = {
        showModal: false,
        showModal2: true,
        showModal3: false
      }

      component = render(
        <Provider store={store}>
          <Router history={history}>
            <QuestionView
              modalState={modalState}
              userData={userData}
              notificationData={notificationData}
              currentAnswer={currentAnswer}
              currentQuestion={currentQuestion}
              quiz={quiz}
              answers={answers}
              showModal={showModal}
              showModal2={showModal2}
              hideModal={hideModal}
              setNotification={setNotification}
              deleteQuestion={deleteQuestion}
              sessionExpired={sessionExpired}
              globals={globals}
            />
          </Router>
        </Provider>
      );
    })


    it("should call action creators to delete a question, set notification and change url", () => {

      const requestResponse = {
        data: "DELETED"
      }

      fireEvent.click(component.getByTestId("modal-delete-button"))
      expect(deleteQuestion).toHaveBeenCalledTimes(1)
    })
  })

  describe("handleDeleteAnswer", () => {

    let component

    beforeEach(() => {
      modalState = {
        showModal: true,
        showModal2: false,
        showModal3: false
      }

      component = render(
        <Provider store={store}>
          <Router history={history}>
            <QuestionView
              modalState={modalState}
              userData={userData}
              notificationData={notificationData}
              currentAnswer={currentAnswer}
              currentQuestion={currentQuestion}
              quiz={quiz}
              answers={answers}
              showModal={showModal}
              showModal2={showModal2}
              hideModal={hideModal}
              setNotification={setNotification}
              deleteAnswer={deleteAnswer}
              sessionExpired={sessionExpired}
              globals={globals}
            />
          </Router>
        </Provider>
      );
    })

    it("should call action creators to delete an answer, set notification and change url", () => {

      const requestResponse = {
        data: "DELETED"
      }

      fireEvent.click(component.getByTestId("modal-delete-button"))
      expect(deleteAnswer).toHaveBeenCalledTimes(1)
    })
  })
})
