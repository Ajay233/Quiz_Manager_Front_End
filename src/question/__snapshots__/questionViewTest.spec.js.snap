// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionView should render the QuestionView component, question details and answers 1`] = `
<div
  className="componentContainer"
  id="questionView"
>
  <a
    className="link back"
    onClick={[Function]}
  >
    <i
      className="fas fa-chevron-circle-left blue"
    />
     Back
  </a>
  <div
    className="questionFormArea"
  >
    <div
      className="questionForm"
    >
      <div
        className="questionViewTitle"
      >
        Question 1 from the Quiz 1 quiz
      </div>
      <div
        className="questionDetailsContainerLarge"
      >
        <div
          className="detailsHeading"
        >
          Question description:
        </div>
        <div
          className="detailsContent"
        >
          Test question
        </div>
      </div>
    </div>
    <div
      className="questionImageArea"
    >
      <div
        className="imgView"
      >
        <img
          alt=""
          className="preview"
        />
      </div>
      <div
        className="link deleteImg"
        onClick={[Function]}
      >
        <span>
          <i
            className="fas fa-trash-alt red"
          />
           Delete Image
        </span>
      </div>
    </div>
  </div>
  <div
    className="headerContainer"
  >
    <div
      id="answersTitle"
    >
      Answers
    </div>
    <a
      className="addButton linkStandard"
      onClick={[Function]}
    >
      <i
        className="fas fa-plus-circle green"
      />
       Add an answer
    </a>
  </div>
  <div
    className="answerHeadings"
    id="answerHeadings"
  >
    <div
      id="numberHeader"
    >
      Answer
    </div>
    <div
      id="descriptionHeader"
    >
      Description
    </div>
    <div
      id="correctAnswerHeading"
    >
      Correct answer?
    </div>
    <div
      id="optionsHeader"
    />
  </div>
  <div
    id="answers"
  >
    <div>
      <div>
        <div
          className="answerContainer"
        >
          <div
            className="ansNumber"
          >
            1
          </div>
          <div
            className="ansDecription"
          >
            Test answer
          </div>
          <div
            className="correctAns"
          >
            <i
              className="far fa-times-circle red"
            />
          </div>
          <div
            className="options"
          >
            <a
              className="deleteOption linkStandard"
              id="deleteAnswer"
              onClick={[Function]}
            >
              <i
                className="fas fa-trash-alt red"
              />
               Delete
            </a>
            <a
              className="edit linkStandard"
              id="editAnswer"
              onClick={[Function]}
            >
              <i
                className="fas fa-edit blue"
              />
               Edit
            </a>
          </div>
        </div>
      </div>
      <div>
        <div
          className="answerContainer"
        >
          <div
            className="ansNumber"
          >
            2
          </div>
          <div
            className="ansDecription"
          >
            Test answer2
          </div>
          <div
            className="correctAns"
          >
            <i
              className="far fa-check-circle green"
            />
          </div>
          <div
            className="options"
          >
            <a
              className="deleteOption linkStandard"
              id="deleteAnswer"
              onClick={[Function]}
            >
              <i
                className="fas fa-trash-alt red"
              />
               Delete
            </a>
            <a
              className="edit linkStandard"
              id="editAnswer"
              onClick={[Function]}
            >
              <i
                className="fas fa-edit blue"
              />
               Edit
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`QuestionView should render the QuestionView component, update question form and no answers if the answers array is empty 1`] = `
<div
  className="componentContainer"
  id="questionView"
>
  <a
    className="link back"
    onClick={[Function]}
  >
    <i
      className="fas fa-chevron-circle-left blue"
    />
     Back
  </a>
  <div
    className="questionFormArea"
  >
    <div
      className="questionForm"
    >
      <div
        className="questionViewTitle"
      >
        Question 1 from the Quiz 1 quiz
      </div>
      <div
        className="questionDetailsContainerLarge"
      >
        <div
          className="detailsHeading"
        >
          Question description:
        </div>
        <div
          className="detailsContent"
        >
          Test question
        </div>
      </div>
    </div>
    <div
      className="questionImageArea"
    >
      <div
        className="imgView"
      >
        <img
          alt=""
          className="preview"
        />
      </div>
      <div
        className="link deleteImg"
        onClick={[Function]}
      >
        <span>
          <i
            className="fas fa-trash-alt red"
          />
           Delete Image
        </span>
      </div>
    </div>
  </div>
  <div
    className="headerContainer"
  >
    <div
      id="answersTitle"
    >
      Answers
    </div>
    <a
      className="addButton linkStandard"
      onClick={[Function]}
    >
      <i
        className="fas fa-plus-circle green"
      />
       Add an answer
    </a>
  </div>
  <div
    className="answerHeadings"
    id="answerHeadings"
  >
    <div
      id="numberHeader"
    >
      Answer
    </div>
    <div
      id="descriptionHeader"
    >
      Description
    </div>
    <div
      id="correctAnswerHeading"
    >
      Correct answer?
    </div>
    <div
      id="optionsHeader"
    />
  </div>
  <div
    id="answers"
  />
</div>
`;
